src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T value = currentNode.value;
+        return value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T lastValue = currentNode.value;
+        currentNode.value = value;
+        return lastValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        unlink(currentNode);
+        T removedValue = currentNode.value;
+        return removedValue;
```

Comment:
```suggestion
        checkIndex(index);
        return unlink(getNodeByIndex(index));
```
Optional suggestion.